include/pybind11/detail/type_caster_base.h

```diff
@@ -28,34 +28,44 @@
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+
 /// A life support system for temporary objects created by `type_caster::load()`.
 /// Adding a patient will keep it alive up until the enclosing function returns.
 class loader_life_support {
 private:
-    void* parent = nullptr;
+    loader_life_support* parent = nullptr;
     std::unordered_set<PyObject *> keep_alive;
 
+    static loader_life_support** get_stack_pp() {
+#if defined(WITH_THREAD)
+        thread_local static loader_life_support* per_thread_stack = nullptr;
```

Comment:
Does this really maintain compatibility though? @wjakob ? Sounds like the better thing to do would be to break the ABI.

Reply:
It does. All the older code is not thread safe, but the newer code is.

I personally think that this is probably better; it's certainly simpler and more conforming to C++ style. If thread_local doesn't work here there are larger problems in the system.

Reply:
@laramiel But is no way to accomplish this with Python's TLS?

Reply:
It can be done, however the python TLS version is much more complicated than this version, and likely performs worse. I don't see why we'd eschew a simple elegant fix in favor of a python api call.

Reply:
Mainly referring to: https://github.com/pybind/pybind11/issues/2765#issuecomment-911846758

Reply:
I think that thread_local is likely to generate code that's at least equivalent or better than by-hand calls to a python thread api with a lot less code, and which the compiler platform can optimize, makes it preferable. Particularly since it doesn't have to be sharable across .so boundaries.

As for the generated code, it's pretty concise, really: https://godbolt.org/z/K57oj9M4P